### ISSUE TITLE
Ctrl-J doesn't add a newline before the command's output

### ISSUE THREAD

**Issue by caribpa** [rizinorg/rizin#612]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Arch Linux x64
| File format of the file you reverse (mandatory)      | n/a
| Architecture/bits of the file (mandatory)            | n/a
| `rizin -v` full output, **not truncated** (mandatory)  | rizin 0.2.0-git @ linux-x86-64, package: 0.2.0.r25723.954ea58ee-git (AUR) commit: 954ea58ee25064fb68d04d06b88ae1599da5154c, build: 2021-02-12__17:02:24

### Expected behavior
Using `Ctrl-J` works like pressing `Ctrl-M` or `Enter`.

`Ctrl-J` is an alternative for `Ctrl-M` and `Enter` in popular terminals (xterm, urxvt), as well as tty/ptys, and some editors (such as graphic emacs, though with some slight differences: `\n` vs `\r`).

### Actual behavior
Using `Ctrl-J` in Rizin doesn't add a newline after the command, making its output override the prompt and the command just written.

### Steps to reproduce the behavior
- `rz --`
- => Presses `Ctrl-J`
- 💔 - No new prompt appears
- `ls` => Use `Ctrl-J` to issue the command
- => Repeat the last `ls` with `Ctrl-J`
- 💔 - Scroll up to see there's no prompt in between the two `ls` executions nor before the first `ls`

### Additional Logs, screenshots, source code,  configuration dump, ...
This issue is also present in radare2.

**Comment by valdaarhun** [comment]
Hi, I would like to work on this issue. I have already gone through several files that I thought would be relevant for this issue. I have also gone through parts of the ebook and doxygen docs. I am not sure if I have found the right location but I think the changes will have to be made in the dietline.c file near lines 1530-1532 and in the `rz_cons_break_pop()` function.
```
case 10: // ^J -- ignore
     rz_cons_break_pop();
     return I.buffer.data;
```

**Comment by valdaarhun** [comment]
There is also one thing that I haven't understood. In `I.buffer.data` (in the previous comment), what does `I` refer to? In the doxygen  docs, when I click on `I` it takes me to this line (line 164 in `arc.h`):
```
 #define I(x) (((unsigned) (x) & 31) << 27)
```
But isn't I(x) like a math expression?

**Comment by ret2libc** [comment]
> There is also one thing that I haven't understood. In `I.buffer.data` (in the previous comment), what does `I` refer to? In the doxygen docs, when I click on `I` it takes me to this line (line 164 in `arc.h`):

I suggest you use some better IDE/tool to navigate the codebase (e.g. clangd, ccls, Visual Studio Code, etc.). It will allow you to go from a use of a variable/symbol to its definition in a more precise way.

Anyway, in the case of dietline.c, `I` is defined in `cons.c` as:
```C
static RzCons rz_cons_instance = { 0 };
#define I rz_cons_instance
```

Thanks for working on this by the way :) Much appreciated.

**Comment by valdaarhun** [comment]
Thanks for the quick response. Will do!

**Comment by valdaarhun** [comment]
Hi. I have been trying to understand the code base (for example, in the dietline.c file) and I feel kind of lost. I have managed to understand a few things in the code but there are many things I haven't been able to understand. The documentation (such as the doxygen docs) only describe the function definitions (return types, arguments, data members in structs, etc) but don't have the actual description of what they do. For example, I haven't understood what the `hud` data member does in the `rz_line_t` struct.

What approach should I take in order to understand the code base well? Are there any docs or sources that I might have overlooked?

**Comment by caribpa** [comment]
@valdaarhun not long ago I started writing an onboarding guide to help newcomers understand the codebase, unfortunately I won't be able to finish it for now due to unfortunate circumstances (maybe I'll have time in 2-3 weeks).

What you can do is ask the experts in the [Rizin official chat](https://im.rizin.re/) for guidance. They are really nice people and would be more than happy to help you 🙂

**Comment by valdaarhun** [comment]
@caribpa Thanks! I'll do that.

**Comment by ret2libc** [comment]
@valdaarhun I've looked a bit at the code... It seems you have found the piece of code that handles `^J`. Now, what I did to understand what's wrong with it, was to look at how `enter` behaves. How does `enter` adds the newline for the prompt? After finding that, it should be easy to find the fix. Hope it helps!

**Comment by valdaarhun** [comment]
@ret2libc Yeah, I found the `enter` block of the code and realised that that should help resolve the issue. But it took me time to understand what each part of the code does. I still haven't understood each part exactly. But given that `Ctrl+J` is supposed to do the same job as `enter`, I thought it would be fine to have `enter`'s code in the `Ctrl-J` block as well.

**Comment by ret2libc** [comment]
@valdaarhun well, understanding everything is quite hard ;) even after years many of us don't understand all details, but we try to understand as needed. It's impossible to understand everything since the project had/has many contributors, so most of the time you touch code written by someone else.

**Comment by valdaarhun** [comment]
@ret2libc Yep. In fact, I had to experiment a lot to find out which part of the code was mainly responsible for the newline. I was continuously commenting, rebuilding, testing and uncommenting. Finally, I realised that in the `_end` section, the block
```
	if (I.echo) {
		printf("\r%s%s\n", I.prompt, I.buffer.data);
		fflush(stdout);
	}
```
was responsible. Because when I commented out everything else except for this it would print the newline, but when I commented this out as well, it wouldn't print it.
